# 3月3日

#### (1) 在郑友捷同学的帮助下完成execve syscall的修复：
发现问题有二：
1.current_task.set_name(name);会访问不正确的地址导致出现缺页异常：
![](../../asserts/day228.png ':class=myImageClass')
![](../../asserts/day303-1.png ':class=myImageClass')

经调试后发现似乎是String::from 的问题，传入的name处于不被支持访问的空间内：
![](../../asserts/day303-2.png ':class=myImageClass')

将current_task.set_name(name);改成current_task.set_name(&program_name);可以避开这个问题
2.aspace被lock后未手动释放，导致出现重复获得锁的错误：
![](../../asserts/day303-3.png ':class=myImageClass')

exec函数会手动改动执行流，执行完 enter_uspace 之后会进入用户态而不返回exec函数，从而导致在函数开头被lock的aspace无法被函数自动释放，因此需要在函数末enter_uspace之前加入drop(aspace);来将锁释放。
至此完成了对syscall execve的支持：
![](../../asserts/day303-4.png ':class=myImageClass')
![](../../asserts/day303-5.png ':class=myImageClass')

#### (2) 与陈羿华同学讨论并完成clone syscall的修复：
前期调试后将问题定位在sysclone调用clone_task时，由 let new_aspace = current_aspace.clone_or_err()?;返回错误：
![](../../asserts/day303-6.png ':class=myImageClass')
![](../../asserts/day303-7.png ':class=myImageClass')

陈羿华同学研究后发现是buffer size过大导致内存分配错误并修复，至此完成了对syscall clone的支持，并因此可支持syscall wait4和syscall sched_yield。

#### basic测例中进程相关的syscall均已通过（除times）：
rv：
![](../../asserts/day303-8.png ':class=myImageClass')
![](../../asserts/day303-9.png ':class=myImageClass')

la：
![](../../asserts/day303-10.png ':class=myImageClass')
![](../../asserts/day303-11.png ':class=myImageClass')